types/decimal.go

```diff
@@ -318,30 +318,74 @@ func (d Dec) QuoInt64(i int64) Dec {
 	return Dec{mul}
 }
 
-// ApproxSqrt returns an approximate sqrt estimation using Newton's method to
-// compute square roots x=√d for d > 0. The algorithm starts with some guess and
+// ApproxRoot returns an approximate estimation of a Dec's positive real nth root
+// using Newton's method (where n is positive). The algorithm starts with some guess and
 // computes the sequence of improved guesses until an answer converges to an
-// approximate answer. It returns -(sqrt(abs(d)) if input is negative.
-func (d Dec) ApproxSqrt() Dec {
+// approximate answer.  It returns -(sqrt(abs(d)) if input is negative.
+func (d Dec) ApproxRoot(root uint64) (guess Dec, err error) {
```

Comment:
Where does the panic stem from? I believe all Dec/Int methods panic when too large. Why have this method deviated from that pattern?

Reply:
It comes from [L353](https://github.com/cosmos/cosmos-sdk/pull/5447/files/b0b39eced9782059f41fa6c6d12ff980644fd8b8#diff-9da3d5f08ea933ebaf3d7d0b4740a346R353) where we take previous `guess`, to the power of `root-1`.  This causes the big Int to overflow.  I feel an error is more proper here, than a panic, because its harder to predict when it will happen.  Like in normal arithmetic operations, you know that your inputs are in the danger zone.  It's a bit harder to predict here

Reply:
I see.

> It comes from L353 where we take previous guess, to the power of root-1. This causes the big Int to overflow.

Is there a way to prevent a guess or short-circuit when we know it'd overflow? This way we can avoid the error return and panic.

Reply:
You mean try to catch the panic, and then restart with a different guess?  That might get into a loop that never ends then?